<a href="https://colab.research.google.com/github/JunwenFang/1015-Text-as-Data/blob/main/hw1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Homework 1 (36 points)

*Due 11:59pm, February 28, 2025 through Brightspace*

Based on material from

- Lectures: Week 1, Week 2, Week 3 (up to Slide 53)
- Lab sessions: Recitations 0, 1, 2

Required files (both on Brightspace)

- wiki40b.txt
- universal_dependencies.txt

As this is the first homework assignment, please be sure to read all the instructions especially carefully.

1. All the work must be your own. You must not copy anyone's work, or allow anyone to copy yours (both the code and written responses). You may discuss the assignment with others, but you must prepare the responses alone.
2. You should 1) create a copy of this notebook, 2) rename it as f"hw1_{your_netid}.ipynb" (e.g. hw1_aqe69420.ipynb), 3) prepare the responses, 4) download it as **.ipynb** (with all code output), and 5) submit it on Brightspace. Not following these formatting instructions greatly slows down our grading speed...
3. Your code must be included in full so that we can assess your understanding of the problems. Note that code comments will not be graded.
4. This notebook contains all the code/text blocks you need to complete the assignment. If you've added/deleted blocks accidentally, just refer back to this original document.


Please see below for how an example question and response might be formatted (you are expected to add your answer after every blockquote character \>).

---


### Example Question

Split the sentence *New York University (NYU) is a private research university in New York City, New York* by whitespace and report how many unique kinds of whitespace-delimited words there are in the sentence (2 points; 1 point for answer + 1 point for code).

In [ ]:
# code for Example Question

def split_string(my_string):
  return set(my_string.split())

nyu = split_string("New York University (NYU) is a private research university in New York City, New York")
print(nyu)
print(len(nyu))

{'in', 'City,', 'New', '(NYU)', 'a', 'research', 'private', 'University', 'university', 'York', 'is'}
11


Answer to Example Question:
> 11

---
#### Begin actual assignment:

## Basic Information

1. Please enter your name (as it appears on Brightspace).
>

2. Who did you collaborate with on this assignment (enter N/A if no one)?
>

3. Did you use generative AI tools, and if so, which tools and for what purpose (enter N/A if none)?
>


## Question 1 (6 points)

Objective: Question 1 tests your understanding of basic definitions.

Given two vectors $$\mathbf{x} = \begin{bmatrix} 1 \\ 3 \\ 5 \\ 7 \end{bmatrix}, \ \mathbf{y} = \begin{bmatrix} 2 \\ 4 \\ 6 \\ 8 \end{bmatrix},$$

1-1. Calculate the Euclidean distance between $\mathbf{x}$ and $\mathbf{y}$ (2 points; 1 point for answer + 1 point for code).

1-2. Calculate the Manhattan distance between $\mathbf{x}$ and $\mathbf{y}$ (2 points; 1 point for answer + 1 point for code).

1-3. Calculate the cosine similarity between $\mathbf{x}$ and $\mathbf{y}$ (2 points; 1 point for answer + 1 point for code).

The use of libraries is **not allowed** for Question 1.

In [13]:
# use this block for 1-1, 1-2, and 1-3
# you can write a function to calculate the quantities,
# or write out the equation and evaluate the answer.

# code for 1-1
# Define vectors x and y
x = [1, 3, 5, 7]
y = [2, 4, 6, 8]

euclidean_distance = ((x[0] - y[0])**2 + (x[1] - y[1])**2 + (x[2] - y[2])**2 + (x[3] - y[3])**2) ** 0.5

# code for 1-2
manhattan_distance = abs(x[0] - y[0]) + abs(x[1] - y[1]) + abs(x[2] - y[2]) + abs(x[3] - y[3])

# code for 1-3
dot_product = x[0]*y[0] + x[1]*y[1] + x[2]*y[2] + x[3]*y[3]
magnitude_x = (x[0]**2 + x[1]**2 + x[2]**2 + x[3]**2) ** 0.5
magnitude_y = (y[0]**2 + y[1]**2 + y[2]**2 + y[3]**2) ** 0.5
cosine_similarity = dot_product / (magnitude_x * magnitude_y)

print(euclidean_distance, manhattan_distance, cosine_similarity)



2.0 4 0.9960238411119947


Answer to 1-1:
> 2

Answer to 1-2:
> 4

Answer to 1-3:
> 0.996

## Question 2 (8 points)

Objective: Question 2 tests your skill to preprocess "real" text data, as might be required for your project. As mentioned in the lecture, text data is often noisy and can require more thought and work to preprocess than textbook examples.

2-1. Preprocess the three Wikipedia articles in the provided "wiki40b.txt" and report the length of the resulting vocabulary, $|V|$. Follow these instructions (4 points; 1 point for answer + 3 points for code):
- Preprocess only the paragraph text (no titles or headers)
- No special tags (like *\_NEWLINE\_*) should be included
- Apply lowercasing
- Remove punctuation

In [14]:
# code for 2-1
# the use of libraries is allowed

from google.colab import files

uploaded = files.upload()  # 运行后选择文件上传

Saving universal_dependencies.txt to universal_dependencies (1).txt
Saving wiki40b.txt to wiki40b (1).txt


In [19]:
file_path = '/content/wiki40b.txt'
with open(file_path, 'r', encoding='utf-8') as file:
  text = file.read()
text

'_START_ARTICLE_\nKasturba Road\n_START_PARAGRAPH_\nKasturba Road is a street in Bangalore, the capital of Karnataka, India, which is connected to M G Road to the north and J C Road to the south. Some important landmarks situated along Kasturba Road are Sree Kan\nteerava Stadium, Kanteerava Indoor Stadium, Cubbon Park, Government Museum, Venkatappa Art Gallery, Visvesvaraya Industrial and Technological Museum and UB City. A 600-year-old Ganesha temple is also situated o\nn Kasturba Road._NEWLINE_It was earlier known as Sydney Road._NEWLINE_Other important landmarks close to the road are Karnataka High Court, Vidhana Soudha and Chinnaswamy Stadium.\n\n_START_ARTICLE_\nAmazon (yacht)\n_START_SECTION_\nConstruction\n_START_PARAGRAPH_\nCarvel planked in teak and pitch pine on oak frames, with alternate wrought iron strap floor reinforcement, bronze fastenings, lead keel and copper sheathing, the Amazon\'s hull is still largely original.\n_START_SECTION_\nHistory\n_START_PARAGRAPH_\nHer bui

In [24]:
import re
import string

# Remove special tags
text = text.replace('_NEWLINE_', ' ')

# Convert to lowercase
text = text.lower()

# Remove punctuation
text = text.translate(str.maketrans('', '', string.punctuation))

# Tokenize words (split by whitespace)
words = text.split()
# Construct vocabulary (unique words)
vocabulary = set(words)

len(vocabulary)


419

Answer to 2-1:
>

2-2. Preprocess the three sentences in the provided "universal_dependencies.txt" and print the resulting vocabulary and report its length. Follow these instructions (4 points; 1 point for answer + 3 points for code):

- Preprocess only the words in the tab-separated columns
- Build a vocabulary by joining the lemma in the third column and the part-of-speech in the fourth column with an underscore (like *underlie\_VERB*)
- No meta information (like *\# sent_id = GUM_academic_exposure-2*) should be included

In [ ]:
# code for 2-2
# the use of libraries is allowed
file_path = '/content/universal_dependencies.txt'
with open(file_path, 'r', encoding='utf-8') as file:
  text = file.read()
text

Answer to 2-2 (just the length, the print output should be part of the code output above):
>

## Question 3 (8 points)

Objective: Question 3 tests your ability to interpret the data correctly and draw the correct conclusions.

3-1. Your zoo enthusiast friend recently visited two zoos (Zoo 1 and Zoo 2) and built an animal-zoo frequency vector (analogous to term-document frequency vector) at each zoo. They share with you the following vocabulary and vector for each zoo:

In [27]:
from sklearn.metrics.pairwise import euclidean_distances

vocab_zoo1 = ["lion", "tiger", "elephant", "giraffe", "zebra", "panda", "kangaroo", "monkey", "wolf", "bear"]
vector_zoo1 = [23, 17, 11, 25, 9, 12, 1, 8, 15, 14]

vocab_zoo2 = ["dolphin", "giraffe", "monkey", "penguin", "crocodile", "bear", "hippopotamus", "shark", "wolf", "panda"]
vector_zoo2 = [17, 3, 2, 15, 22, 1, 8, 12, 9, 3]

print(euclidean_distances([vector_zoo1, vector_zoo2]))

[[ 0.         30.41381265]
 [30.41381265  0.        ]]


Based on the output above, they conclude that the Euclidean distance of the two animal-zoo frequency vectors is about 30.41. Why is their reasoning incorrect (2 points)?
>

> Because two zoo frequency vectors have different vocabularies. The Euclidean distance calculation should be performed only on the common vocabulary, while for missing animals, their frequency should be treated as 0. They should use cosine similarity.

3-2. Calculate the correct Euclidean distance between the two vectors (2 points; 1 point for answer + 1 points for code).

In [28]:
# code for 3-2
# the use of libraries is allowed
# you can assume that the variables from the code block above carries over

import numpy as np

# Build the unified vocabulary
all_animals = sorted(set(vocab_zoo1 + vocab_zoo2))  # Sorted for consistency

# Create aligned frequency vectors
vector1_aligned = [vector_zoo1[vocab_zoo1.index(animal)] if animal in vocab_zoo1 else 0 for animal in all_animals]
vector2_aligned = [vector_zoo2[vocab_zoo2.index(animal)] if animal in vocab_zoo2 else 0 for animal in all_animals]

# Compute the Euclidean distance
euclidean_distance = np.sqrt(sum((np.array(vector1_aligned) - np.array(vector2_aligned)) ** 2))
print("Correct Euclidean distance:", euclidean_distance)


Correct Euclidean distance: 55.072679252057455


Answer to 3-2:
> 55.07

3-3. Which three animals 'contribute most' to the Euclidean distance between the two vectors (which three animals have the biggest difference in frequency; 2 points; 1 point for answer + 1 points for code)?

In [35]:
# code for 3-3
# the use of libraries is allowed
# you can assume that the variables from the code block above carries over
# Compute absolute differences
diffs = np.abs(np.array(vector1_aligned) - np.array(vector2_aligned))

# Get the indices of the top 3 contributors
top_3_indices = np.argsort(diffs)[-3:]

# Print the three animals with the highest contribution
top_3_animals = [all_animals[i] for i in top_3_indices]
print("Top 3 contributing animals:", top_3_animals)


Top 3 contributing animals: ['crocodile', 'giraffe', 'lion']


Answer to 3-3:
> * crocodile
* giraffe
* lion

3-4. Based on your responses to 3-2 and 3-3, what reasonable conclusions can you draw about the difference between Zoo 1 and Zoo 2? Assume your friend correctly counted all the animals at the two Zoos (answer in 1-2 sentences; 2 points).
> The difference between Zoo1 and Zoo2 is actually larger than my friend thinks. Also, the main difference doesn't come from the difference of species in two zoos. Instead, the number of animals such as crocodiles, giraffe, lions causes the main discrepancy.

>

## Question 4 (6 points)

Objective: Question 4 is about co-occurrence matrices, which are important for text analysis and NLP.

4-1. Assume that the predefined `split_data` variable below is a sequence of preprocessed words. Using this data and the provided `word_index`, build and print two co-occurrence matrices using different context windows (4 points; 1 point for print output + 3 points for code).
- Matrix 1: The context window is one word to the **right**.
- Matrix 2: The context window is one word to the **left**.

Hint: Adapt the code from Recitation #2, but be careful, because this is different from the lab session example, in which the context window was to both the left and the right.

Warning: As `["A", "B", "C", "D"]` occurs as both the word and the context, it may be easy to confuse what the rows and columns represent in the matrices. Not confusing this is important for 4-2 below.

In [ ]:
# build on the code below for 4-1
# the use of libraries is allowed
# use these pre-defined variables below
raw_data = "D D A B C D D A B D D D A B C D A B C A B C C C A B A B D A B D C A B C A B A B D C C D A B D A B A B D D D D C D C D A B A B A B D A B C C C C A B D C C C C D D D C A B C D A B C A B A B A B C D A B C D D C A B D A B D D A B A B C C A B C C D A B D A B A B C D D A B D C D A B D D A B D D C A B D A B C A B C D A B A B A B D A B C D A B D A B C D A B D A B C A B A B A B D A B D C A B D A B C D A B A B A B A B C D C A B C A B D D A B C D D C D C D D D D C C A B A B C D A B A B C C A B A B C A B D D D A B D C C D D D A B C C D C D A B C D D D D D D C C D C A B C C C D C D C A B A B A B D C D D A B A B A B D C A B C D D C A B A B A B A B D A B A B A B C C D A B C C C D D A B D C D A B A B A B D A B A B A B C A B D D C C C A B A B D A B D D D C D D D C A B D A B D A B C C A B C A B C C D A B A B D C A B D C A B D D C D D A B A B C D C D D C C C D D A B C A B D C C C C D A B A B A B C D C C D D C D D C A B A B D C A B D D D D C C D D A B D D C D C D C A B A B D A B D C D A B D A B A B C D A B D C A B C D D A B A B D D A B C C C C D C D D C A B C C D C A B D A B A B A B A B D A B C C C D D C C C D D D A B A B D D D D D C C C D A B D D C A B D A B D D C A B C D D D C D A B D C A B D C D C A B C A B A B D D D D D D C D D D D D D C C C D D C A B D C C D A B A B D A B D C A B C C A B A B A B A B C C"
split_data = raw_data.split()
word_index = {word: idx for idx, word in enumerate(["A", "B", "C", "D"])}

print(len(split_data))  # 655



4-2. Upon visual inspection of the matrices (you don't need to calculate any distances), are the representations for each word the same or different across the two matrices? Why or why not (2 points; 1 point for correctly identifying same/different + 1 point for reason)?
>

## Question 5 (8 points)

Objective: Question 5 is to prepare you for the final research project.

5-1. Formulate a research question you would like to tackle for your final project. This will be evaluated in light of the FINER method introduced in the first lecture (2 points).
> How do Federal Reserve statements and speeches influence market sentiment and liquidity, and can we develop an NLP-based predictive model to quantify these effects in the change of SPY price next period.

5-2. In about 2-3 sentences, convince us why we should care about this research question. Can it reveal new scientific findings about the world? Can it lead to useful engineering applications (3 points)?
> Understanding the impact of Federal Reserve communications on financial markets is crucial for policymakers, investors, and economists. By analyzing the language and tone of Fed statements, we can uncover hidden sentiment trends that drive market movements. This research can lead to practical applications in trading and economic forecasting by providing an early warning system for market fluctuations.


5-3. In about 2-3 sentences, describe why this research question is best answered using text data, and what kind of text data you expect to use (3 points).
> Because instead of certain numerical data, Federal Reserve statements, speeches, and meeting contain qualitative insights into economic outlooks and policy intentions. The data sources will include transcripts of FOMC meetings, official Fed communications, and financial news articles. By applying NLP techniques such as sentiment analysis, topic modeling, and text embeddings, we can quantify the impact of Fed communications on financial markets.